packages/flutter/lib/src/services/text_editing_value.dart

```diff
@@ -0,0 +1,444 @@
+    return TextEditingValue(
+      text: encoded['text'] as String,
+      selection: TextSelection(
+        baseOffset: encoded['selectionBase'] as int? ?? -1,
+        extentOffset: encoded['selectionExtent'] as int? ?? -1,
+        affinity: _toTextAffinity(encoded['selectionAffinity'] as String?) ?? TextAffinity.downstream,
+        isDirectional: encoded['selectionIsDirectional'] as bool? ?? false,
+      ),
+      composing: TextRange(
+        start: encoded['composingBase'] as int? ?? -1,
+        end: encoded['composingExtent'] as int? ?? -1,
+      ),
+    );
+  }
+
+  /// The current text being edited.
+  final String text;
+
+  /// The range of text that is currently selected.
+  final TextSelection selection;
+
+  /// The range of text that is still being composed.
+  final TextRange composing;
+
+  /// A value that corresponds to the empty string with no selection and no composing range.
+  static const TextEditingValue empty = TextEditingValue();
+
+  // TODO(gspencergoog): replace when we expose this ICU information.
+  /// Check if the given code unit is a white space or separator
+  /// character.
+  ///
+  /// Includes newline characters from ASCII and separators from the
+  /// [unicode separator category](https://www.compart.com/en/unicode/category/Zs)
+  static bool isWhitespace(int codeUnit) {
+    switch (codeUnit) {
+      case 0x9: // horizontal tab
+      case 0xA: // line feed
+      case 0xB: // vertical tab
+      case 0xC: // form feed
+      case 0xD: // carriage return
+      case 0x1C: // file separator
+      case 0x1D: // group separator
+      case 0x1E: // record separator
+      case 0x1F: // unit separator
+      case 0x20: // space
+      case 0xA0: // no-break space
+      case 0x1680: // ogham space mark
+      case 0x2000: // en quad
+      case 0x2001: // em quad
+      case 0x2002: // en space
+      case 0x2003: // em space
+      case 0x2004: // three-per-em space
+      case 0x2005: // four-er-em space
+      case 0x2006: // six-per-em space
+      case 0x2007: // figure space
+      case 0x2008: // punctuation space
+      case 0x2009: // thin space
+      case 0x200A: // hair space
+      case 0x202F: // narrow no-break space
+      case 0x205F: // medium mathematical space
+      case 0x3000: // ideographic space
+        break;
+      default:
+        return false;
+    }
+    return true;
+  }
+
+  /// Returns the index into the string of the next character boundary after the
+  /// given index.
+  ///
+  /// The character boundary is determined by the characters package, so
+  /// surrogate pairs and extended grapheme clusters are considered.
+  ///
+  /// The index must be between 0 and string.length, inclusive. If given
+  /// string.length, string.length is returned.
+  ///
+  /// Setting includeWhitespace to false will only return the index of non-space
+  /// characters.
+  static int nextCharacter(int index, String string, [bool includeWhitespace = true]) {
```

Comment:
I have written similar typographic methods on the engine side
https://github.com/flutter/engine/blob/e0ebbb226f38e552289e380324943ec67126d325/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L1192
and
https://github.com/flutter/engine/blob/e0ebbb226f38e552289e380324943ec67126d325/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm#L437

I hope at some point we can use the API from skia to calculate code unit and avoid duplicating the logic in a bunch of places...

Reply:
Ah good catch.  I remember talking to the Dart team about this but can't remember the details.  This seems to work with my testing.

I'll refactor this to use CharacterRange, move it into test_editing_action_target.dart, and make it `visibleForTesting` only.

Reply:
Here's that change: bd27952